octodns/provider/rackspace.py

```diff
@@ -0,0 +1,385 @@
+        create_out = self._create_given_change_values(change, create_values)
+
+        update_out = []
+        update_values = set(new_values).intersection(set(existing_values))
+        for value in update_values:
+            transformer = getattr(self,
+                                  "_record_for_{}".format(change.new._type))
+            prior_rs_record = transformer(change.existing, value)
+            prior_key = self._key_for_record(prior_rs_record)
+            next_rs_record = transformer(change.new, value)
+            next_key = self._key_for_record(next_rs_record)
+            next_rs_record["id"] = self._id_map[prior_key]
+            del next_rs_record["type"]
+            update_out.append(next_rs_record)
+            self._id_map[next_key] = self._id_map[prior_key]
+            del self._id_map[prior_key]
+        return create_out, update_out, delete_out
+
+    def _mod_Delete(self, change):
+        return self._delete_given_change_values(change, self._get_values(
+            change.existing))
+
+    def _delete_given_change_values(self, change, values):
+        out = []
+        for value in values:
+            transformer = getattr(self, "_record_for_{}".format(
+                change.existing._type))
+            rs_record = transformer(change.existing, value)
+            key = self._key_for_record(rs_record)
+            out.append('id=' + self._id_map[key])
+            del self._id_map[key]
+        return out
+
+    def _apply(self, plan):
+        desired = plan.desired
+        changes = plan.changes
+        self.log.debug('_apply: zone=%s, len(changes)=%d', desired.name,
+                       len(changes))
+
+        domain_id = self._get_zone_id_for(desired)
+        creates = []
+        updates = []
+        deletes = []
+        for change in changes:
+            if change.__class__.__name__ == 'Create':
+                creates += self._mod_Create(change)
+            elif change.__class__.__name__ == 'Update':
+                add_creates, add_updates, add_deletes = self._mod_Update(
+                    change)
+                creates += add_creates
+                updates += add_updates
+                deletes += add_deletes
+            else:
+                assert change.__class__.__name__ == 'Delete'
+                deletes += self._mod_Delete(change)
+
+        if deletes:
+            params = "&".join(sorted(deletes))
+            self._delete('domains/{}/records?{}'.format(domain_id, params))
+
+        if updates:
+            data = {"records": sorted(updates, key=lambda v: v['name'])}
+            self._put('domains/{}/records'.format(domain_id), data=data)
+
+        if creates:
+            data = {"records": sorted(creates, key=lambda v: v['type'] +
+                                      v['name'] +
+                                      v.get('data', ''))}
+            self._post('domains/{}/records'.format(domain_id), data=data)
```

Comment:
Are the creates, updates, and deletes broken out into 3 requests due to constraints or could they all be sent in a single request? 

It can be safer to make all the changes in one go if the API supports it since you could be deleting an A record to replace it with a CNAME and if the delete succeeds, but the update doesn't things would be in a "broken" state where neither record would exist. This is a pretty common flaw in the DNS APIs that either don't support batch updates (doesn't appear to be applicable here) or don't correctly do validations on the result of the whole batch. Just want to make sure this is doing things as safely as reasonably possible. I've been slogging through the same mess with the Cloudflare provider in https://github.com/github/octodns/pull/162. Some providers are "safer" than others in this respect.

Separately are there any (obvious/published) size limits to the request data size?

Reply:
RackSpace has a REST API that treats individual record-set entries as resources. To say that this makes for an absolutely *terrible* DNS API is a dramatic understatement. In this specific case, please note that delete, update, and create require DELETE, PUT, and POST HTTP methods, respectively, so simply cannot be done in a single request. Combine this with the one-address-per-"record" rule and the fact that they want their internal `id` for most operations instead of the record data itself and you end up with the nightmare fuel  in _mod_Update. I think this went through 4 or 5 refactors before we finally got all of the constraints addressed well enough for update to work reliably.

Speaking of safety, I'd settle for merely correct. Take a look at the uses of {add,remove}_trailing_dot. The RackSpace API does not allow you to have a trailing dot in the request at all for most (but not all) record types, however internally treats most records as if they had the trailing dot.

It took a lot of trial and error to figure out what their API even wanted in most cases. I think we ended up doing fairly exhaustive testing around the various supported record types -- most of the test cases were built from actual failures we encountered in our own records or in tests we built to explore these edge cases.

Reply:
Sorry, I didn't catch the delete/put/post bit. Makes sense given the constraints and is inline with some of the other API's. If you don't mind just drop a comment in before the 3 calls with a quick bit about why it's that way, how it's not particularly safe, but making the best of things etc.  Mostly so that if others see it in the future or run into problems it'll be called out in place.

Reply:
Also, there do not appear to be size limits on individual requests; all of their throttling appears to be around number and frequency of requests.